common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/YubiKitCertBasedAuthManager.java

```diff
@@ -0,0 +1,226 @@
+    private static final String MDEVICE_NULL_ERROR_MESSAGE = "Instance UsbYubiKitDevice variable (mDevice) is null.";
+    private static final String YUBIKEY_PROVIDER = "YKPiv";
+
+    private final YubiKitManager mYubiKitManager;
+    private UsbYubiKeyDevice mDevice;
+    //Lock to help facilitate synchronization
+    private static final Object sDeviceLock = new Object();
+
+    /**
+     * Create new instance of YubiKitCertBasedAuthManager.
+     * @param context current application context.
+     */
+    YubiKitCertBasedAuthManager(@NonNull final Context context) {
+        mYubiKitManager = new YubiKitManager(context);
+    }
+
+    /**
+     * Create and start YubiKitManager for UsbDiscovery mode.
+     * When in Usb Discovery mode, Yubikeys that plug into the device will be accessible
+     *  once the user provides permission via the Android permission dialog.
+     * @param startDiscoveryCallback Contains callbacks to run when a YubiKey is connected and disconnected.
+     */
+    @Override
+    public void startDiscovery(@NonNull final IStartDiscoveryCallback startDiscoveryCallback) {
+        mYubiKitManager.startUsbDiscovery(new UsbConfiguration(), new Callback<UsbYubiKeyDevice>() {
+            @Override
+            public void invoke(@NonNull UsbYubiKeyDevice device) {
+                Logger.verbose(TAG, "A YubiKey device was connected");
+                synchronized (sDeviceLock) {
+                    mDevice = device;
+                    startDiscoveryCallback.onCreateConnection();
+
+                    mDevice.setOnClosed(new Runnable() {
+                        @Override
+                        public void run() {
+                            Logger.verbose(TAG, "A YubiKey device was disconnected");
+                            synchronized (sDeviceLock) {
+                                mDevice = null;
+                            }
+                            final PivProviderStatusEvent pivProviderStatusEvent = new PivProviderStatusEvent();
+                            //Remove the YKPiv security provider if it was added.
+                            if (Security.getProvider(YUBIKEY_PROVIDER) != null) {
+                                Security.removeProvider(YUBIKEY_PROVIDER);
+                                Telemetry.emit(pivProviderStatusEvent.putPivProviderRemoved(true));
+                                Logger.info(TAG, "An instance of PivProvider was removed from Security static list upon YubiKey device connection being closed.");
+                            } else {
+                                Telemetry.emit(pivProviderStatusEvent.putPivProviderRemoved(false));
+                                Logger.info(TAG, "An instance of PivProvider was not present in Security static list upon YubiKey device connection being closed.");
+                            }
+                            startDiscoveryCallback.onClosedConnection();
```

Comment:
Supposing we get rid of onCreateConnection(), this is what would happen in the "unplug early" scenario:
1. User unplugs in the middle of the dialog flow, so another error dialog pops up asking the user to plug in again.
2. User plugs in their key, and they 'ok' the permission dialog.
3. They still see the error dialog asking them to plug in. 

It's possible to still have a dialog left over upon a new connection when the previous connection ended unexpectedly like this. I think it does result in a better user experience to have the preexisting dialog dismissed, but if the messiness of having onCreateConnection() outweigh this, I'm open to getting rid of it.

IStartDiscoveryCallback also has an additional method onException() that was added in response to a suggestion in this PR. It's not seen in the YubiKitCertBasedAuthManager because YubiKit's discovery methods aren't throwing exceptions, but as Brian pointed out, other types of smartcards could.

I think what I'll do for now is create a variable to hold an instance of IStartDiscoveryCallback in ISmartcardCertBasedAuthManager, and have it set within the constructor. StartDiscovery() then will have no parameters. Let me know if that would work.

Reply:
As for the dialog logic, I think the one way I can see there being issues is if a user tried to tap another smartcard via NFC while another smartcard is currently plugged in. I may need to add some logic then that makes sure every other manager says no device is connected before continuing with a new connection.

Reply:
I thought we would fail the flow if the user unplugs. Isn't that the case?

Reply:
(If we are going to support NFC... 

Given the nature of it (not physically plugged in, so I assume the connection might not be as stable, and intermittent connection loss might happen as the device moves).....  we might need another dialog to let the user explicitly choose which cert/device  will be used to do auth, and then once selected, then we go all the way with that cert/device. )

Reply:
We do fail the cba flow immediately. Any time an error dialog is showing, the request has been cancelled and a server error message shows in the background. 
usb discovery is active for the duration of the webview fragment, so a user is able to replugin their key right away while seeing an error message.

Reply:
Addressing NFC, I was thinking that we could still build upon the 'already plugged in' work by making the criteria of choosing on-device (not having smartcard plugged in) be another dialog prompting the user to either tap their smartcard or plug it in (and then a button at the bottom if the user wants to goto on-device certs). I was hoping to get some cxp input on how they envisioned NFC to work, but I think at this point I'm just going to try a solution out and see what they think.
Since NFC discovery works a bit differently (probably will need to make the user tap twice for reading certs and then checking PIN), this will probably result in another larger feature PR. :^)